qiita_db/sql_connection.py

```diff
@@ -633,3 +581,239 @@ def get_temp_queue(self):
         self.create_queue(temp_queue_name)
 
         return temp_queue_name
+
+
+class Transaction(object):
+    """A context manager that encapsulates a DB transaction
+
+    A transaction is defined by a series of consecutive queries that need to
+    be applied to the database as a single block.
+
+    Parameters
+    ----------
+    name : str
+        Name of the transaction.
+    """
+
+    _regex = re.compile("^{(\d+):(\d+):(\d+)}$")
+
+    def __init__(self, name):
+        # The name is useful for debugging, since we can identify the
+        # failed queue in errors
+        self._name = name
+        self._queries = []
+        self._results = []
+        self._index = 0
+        self._conn_handler = SQLConnectionHandler()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        status = self._conn_handler._connection.get_transaction_status()
+        if exc_type is not None or status != TRANSACTION_STATUS_IDLE:
+            # An exception occurred during the execution of the transaction
+            # Make sure that we leave the DB w/o any modification
+            self.rollback()
+
+    def _raise_execution_error(self, sql, sql_args, error):
+        """Rollbacks the current transaction and raises a useful error
+
+        The error message contains the name of the transaction, the failed
+        query, the arguments of the failed query and the error generated.
+
+        Raises
+        ------
+        ValueError
+        """
+        self.rollback()
+        raise ValueError(
```

Comment:
Can this be a PostgresError? ValueError doesn't fit to me, since it can be a malformed query or other non-value based error. Maybe RuntimeError if Postgres is not doable?